synapse/rest/client/v1/login.py

```diff
@@ -100,6 +120,42 @@ def do_password_login(self, login_submission):
 
         defer.returnValue((200, result))
 
+    @defer.inlineCallbacks
+    def do_cas_login(self, cas_response_body):
+        root = ET.fromstring(cas_response_body)
+        if not root.tag.endswith("serviceResponse"):
+            raise LoginError(401, "Invalid CAS response", errcode=Codes.UNAUTHORIZED)
+        if not root[0].tag.endswith("authenticationSuccess"):
+            raise LoginError(401, "Unsuccessful CAS response", errcode=Codes.UNAUTHORIZED)
+        for child in root[0]:
+            if child.tag.endswith("user"):
+                user = child.text
+                user_id = "@%s:%s" % (user, self.servername)
+                auth_handler = self.handlers.auth_handler
+                user_exists = yield auth_handler.does_user_exist(user_id)
+                if user_exists:
+                    user_id, access_token, refresh_token = yield
+                    auth_handler.login_with_cas_user_id(user_id)
+                    result = {
+                        "user_id": user_id,  # may have changed
+                        "access_token": access_token,
+                        "refresh_token": refresh_token,
+                        "home_server": self.hs.hostname,
+                    }
+
+                else:
+                    user_id, access_token = yield
+                    self.handlers.registration_handler.register(localpart=user)
+                    result = {
+                        "user_id": user_id,  # may have changed
+                        "access_token": access_token,
+                        "home_server": self.hs.hostname,
+                    }
+
+            defer.returnValue((200, result))
```

Comment:
Is it expected that if `child.tag.endswith("user") is False` we no-op and return `200`?